flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/AbstractStateChangeLogger.java

```diff
@@ -145,9 +145,17 @@ protected void log(
             @Nullable ThrowingConsumer<DataOutputViewStreamWrapper, IOException> dataWriter,
             Ns ns)
             throws IOException {
+        log(op, dataWriter, ns, keyContext.getCurrentKeyGroupIndex());
+    }
+
+    protected void log(
+            StateChangeOperation op,
+            @Nullable ThrowingConsumer<DataOutputViewStreamWrapper, IOException> dataWriter,
+            Ns ns,
+            int keyGroup)
```

Comment:
I think the root cause is that we don't have the correct `currentKey` during pqState#poll, if we could modify `InternalTimerServiceImpl` like below, to set the current key before calling poll:

~~~ java
    private void onProcessingTime(long time) throws Exception {
        // null out the timer in case the Triggerable calls registerProcessingTimeTimer()
        // inside the callback.
        nextTimer = null;

        InternalTimer<K, N> timer;

        while ((timer = processingTimeTimersQueue.peek()) != null && timer.getTimestamp() <= time) {
            keyContext.setCurrentKey(timer.getKey());
            processingTimeTimersQueue.poll();
            triggerTarget.onProcessingTime(timer);
        }

        if (timer != null && nextTimer == null) {
            nextTimer =
                    processingTimeService.registerTimer(
                            timer.getTimestamp(), this::onProcessingTime);
        }
    }
~~~

and 

~~~ java
    public void advanceWatermark(long time) throws Exception {
        currentWatermark = time;

        InternalTimer<K, N> timer;

        while ((timer = eventTimeTimersQueue.peek()) != null && timer.getTimestamp() <= time) {
            keyContext.setCurrentKey(timer.getKey());
            eventTimeTimersQueue.poll();
            triggerTarget.onEventTime(timer);
        }
    
~~~

The test could also pass. I think this change looks better with less change and avoid to compute the key group again.

Reply:
Indeed, your proposal is much more simple. However, I think it has several drawbacks:
1. There could be other places where `eventTimeTimersQueue.poll()` or `remove` is called. So it must be a (javadoc) contract that `setCurrentKey` is required before `poll`
2. Following the above contract, `setCurrentKey` should also be called on recovery (by `PriorityQueueStateChangeApplier` - similar to `KvStateChangeApplier`)

Reply:
> 1. There could be other places where eventTimeTimersQueue.poll() or remove is called. So it must be a (javadoc) contract that setCurrentKey is required before poll

As `processingTimeTimersQueue` is only used internally, I think refining the doc and description is okay .

> 2. Following the above contract, setCurrentKey should also be called on recovery (by PriorityQueueStateChangeApplier - similar to KvStateChangeApplier)

I think we don't need to set the current key on recovery, the root cause of this problem is that it did not record the correct key group on logging. And we actually could filter the keygroup out of current range in [condition check](https://github.com/apache/flink/blob/a0d31c5e0914d8e759917a72ca7b667d3db2f1d2/flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/restore/ChangelogBackendLogApplier.java#L95). However, as it did not record the corret key group, some wrong data could pass the filter then.

Reply:
> the root cause of this problem is that it did not record the correct key group on logging

Yes. I mean that such a contract should also be followed in all places.

Reply:
I've opened #19096 to implement this approach.